## Faster R-CNN

Faster R-CNN[1]の実装を動かします。

[1]Ren, S., He, K., Girshick, R., Sun, J.: Faster r-cnn: Towards real-time object detection with region proposal networks. In: Advances in neural information processing systems. (2015) 91–99

[https://arxiv.org/pdf/1506.01497.pdf](https://arxiv.org/pdf/1506.01497.pdf)

以下のものを使用してください。Kerasを使用した実装です。

[duckrabbits/ObjectDetection at master](https://github.com/duckrabbits/ObjectDetection/tree/master)

### 【問題1】学習と推定
READMEを参考に上記実装を動かしてください。

In [ ]:
!pip install tensorflow==1.13.1

In [ ]:
import tensorflow

/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:529: FutureWarning: Passing (type, 1) or 

In [ ]:
tensorflow.__version__

'1.13.1'

In [ ]:
!pip install kaggle

In [ ]:
# 自分のマイドライブにマウントする

from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [ ]:
mkdir .kaggle

In [ ]:
ls -a

./  ../  .config/  drive/  .kaggle/  sample_data/


In [ ]:
#cd drive/"My Drive"

/content/drive/My Drive


In [ ]:
# https://www.kaggle.com/　へアクセスし、ログイン後に右上のプロフィール画像をクリック、さらに"My Account"をクリックする
# このサイトの「API」項目にある"Create New API Token"をクリックすると、kaggle.jsonファイルが自動的にダウンロードされる
# ローカルで、kaggle.json（ダウンロードフォルダにあるはず）をエディターで開く。
# このセルの以下のコードにある token = {'username':'***','key':'***'} における「***」部分を、
# ダウンロードしたkaggle.jsonを参照して書き換え、このセルを実行する

import json

token = {"username":"xxx","key":"xxx"} # マスク
with open('/content/.kaggle/kaggle.json', 'w') as file:
    json.dump(token, file)

In [ ]:
!chmod 600 /content/.kaggle/kaggle.json

In [ ]:
pwd

'/content'

In [ ]:
ls -a .kaggle

./  ../  kaggle.json


In [ ]:
mkdir　/root/.kaggle/

In [ ]:
!cp -R /content/.kaggle/kaggle.json /root/.kaggle/

In [ ]:
!ls -a /root/.kaggle/

.  ..  kaggle.json


In [ ]:
# kaggle APIコマンドが使えるかどうかを確認


!kaggle competitions list

ref                                            deadline             category            reward  teamCount  userHasEntered  
---------------------------------------------  -------------------  ---------------  ---------  ---------  --------------  
contradictory-my-dear-watson                   2030-07-01 23:59:00  Getting Started     Prizes        281           False  
gan-getting-started                            2030-07-01 23:59:00  Getting Started     Prizes        118           False  
tpu-getting-started                            2030-06-03 23:59:00  Getting Started  Knowledge        370           False  
digit-recognizer                               2030-01-01 00:00:00  Getting Started  Knowledge       2430           False  
titanic                                        2030-01-01 00:00:00  Getting Started  Knowledge      18507           False  
house-prices-advanced-regression-techniques    2030-01-01 00:00:00  Getting Started  Knowledge       4789            True  
connectx

In [ ]:
# カレントディレクトリの変更

import os

os.chdir('/content/drive/My Drive/')

# カレントディレクトリの取得
print(os.getcwd())

/content/drive/My Drive


In [ ]:
# simpsonデータセットのAPIコマンドでデータセットをダウンロードする
# APIコマンドはこちらのサイトを参照。https://www.kaggle.com/alexattia/the-simpsons-characters-dataset

!kaggle datasets download -d alexattia/the-simpsons-characters-dataset --force

100% 1.08G/1.08G [00:26<00:00, 79.7MB/s]
100% 1.08G/1.08G [00:26<00:00, 43.0MB/s]


In [ ]:
!unzip /content/drive/'My Drive'/the-simpsons-characters-dataset.zip  -d Simpsons

ストリーミング出力は最後の 5000 行に切り捨てられました。
  inflating: Simpsons/simpsons_dataset/simpsons_dataset/moe_szyslak/pic_1271.jpg  
  inflating: Simpsons/simpsons_dataset/simpsons_dataset/moe_szyslak/pic_1272.jpg  
  inflating: Simpsons/simpsons_dataset/simpsons_dataset/moe_szyslak/pic_1273.jpg  
  inflating: Simpsons/simpsons_dataset/simpsons_dataset/moe_szyslak/pic_1274.jpg  
  inflating: Simpsons/simpsons_dataset/simpsons_dataset/moe_szyslak/pic_1275.jpg  
  inflating: Simpsons/simpsons_dataset/simpsons_dataset/moe_szyslak/pic_1276.jpg  
  inflating: Simpsons/simpsons_dataset/simpsons_dataset/moe_szyslak/pic_1277.jpg  
  inflating: Simpsons/simpsons_dataset/simpsons_dataset/moe_szyslak/pic_1278.jpg  
  inflating: Simpsons/simpsons_dataset/simpsons_dataset/moe_szyslak/pic_1279.jpg  
  inflating: Simpsons/simpsons_dataset/simpsons_dataset/moe_szyslak/pic_1280.jpg  
  inflating: Simpsons/simpsons_dataset/simpsons_dataset/moe_szyslak/pic_1281.jpg  
  inflating: Simpsons/simpsons_dataset/simpsons_dataset

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
!git clone https://github.com/duckrabbits/ObjectDetection.git

Cloning into 'ObjectDetection'...
remote: Enumerating objects: 18, done.
remote: Counting objects: 100% (18/18), done.
remote: Compressing objects: 100% (17/17), done.
remote: Total 46 (delta 3), reused 6 (delta 1), pack-reused 28
Unpacking objects: 100% (46/46), done.


In [ ]:
# https://github.com/duckrabbits/ObjectDetection/tree/master
# この後、上述のgithubのreadmeに沿って、simpsons_datasetフォルダをプロジェクトフォルダ配下に格納しましょう。
# testデータなども同様に準備してみよう。

In [ ]:
mv Simpsons/simpsons_dataset ObjectDetection/

In [ ]:
!mv 

mv: missing file operand
Try 'mv --help' for more information.


In [ ]:
pwd

'/content'

In [ ]:
cd drive/My\ Drive/ObjectDetection

/content/drive/My Drive/ObjectDetection


In [ ]:
pwd

'/content/drive/My Drive/ObjectDetection'

In [ ]:
!pip install keras==2.2.5

     |████████████████████████████████| 337kB 3.4MB/s eta 0:00:01
  Found existing installation: Keras 2.4.3
    Uninstalling Keras-2.4.3:
      Successfully uninstalled Keras-2.4.3


In [ ]:
import keras
print(keras.__version__)

2.2.5


Using TensorFlow backend.


In [ ]:
!python train.py -p annotation.txt

Using TensorFlow backend.
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:529: FutureWarn

In [ ]:
!python predict.py -i kaggle_simpson_testset -c ./save/train_20200925-005356_config.pickle

Using TensorFlow backend.
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:529: FutureWarn

【問題2】コードリーディング
上記実装のコードリーディングを行ってください。

まず、Faster R-CNN[1]において重要だと考えた部分を列挙してください。そして、それに対応するコードを見つけてください。

（例）

- RPNを実現しているコードはどこか \\
model/resnet.pyのrpn関数

- RoIプーリングを実現しているコードはどこか \\
model/RoiPoolingConv.pyのRoiPoolingConvクラス

## YOLOv3

シンプソンズのデータセットをFaster R-CNN以外の手法で学習・推定を行います。YOLOv3[2]のKeras実装を使います。

[qqwweee/keras-yolo3: A Keras implementation of YOLOv3 (Tensorflow backend)](https://github.com/qqwweee/keras-yolo3)

[2]Jeseph Redmon, Ali Farhadi. YOLOv3: An Incremental Improvement

[https://pjreddie.com/media/files/papers/YOLOv3.pdf](https://pjreddie.com/media/files/papers/YOLOv3.pdf)

### 【問題3】学習済みの重みによる推定
学習済みの重みを使い推定を行う方法がREADME.mdのQuick Startの項に記載されています。

まずはこの通りにして各自何かしらの画像や動画に対して検出を行ってください。

出力結果を課題の一部として提出してください。

In [ ]:
cd ..

/content/drive/My Drive


In [ ]:
pwd

'/content/drive/My Drive'

In [ ]:
!git clone https://github.com/qqwweee/keras-yolo3.git

Cloning into 'keras-yolo3'...
remote: Enumerating objects: 144, done.
remote: Total 144 (delta 0), reused 0 (delta 0), pack-reused 144
Receiving objects: 100% (144/144), 151.07 KiB | 2.80 MiB/s, done.
Resolving deltas: 100% (65/65), done.


In [ ]:
cd keras-yolo3/

/content/drive/My Drive/keras-yolo3


In [ ]:
pwd

'/content/drive/My Drive/keras-yolo3'

In [ ]:
!git clone https://github.com/pjreddie/darknet

Cloning into 'darknet'...
remote: Enumerating objects: 5913, done.
remote: Total 5913 (delta 0), reused 0 (delta 0), pack-reused 5913
Receiving objects: 100% (5913/5913), 6.34 MiB | 7.25 MiB/s, done.
Resolving deltas: 100% (3917/3917), done.
Checking out files: 100% (979/979), done.


In [ ]:
cd darknet

/content/drive/My Drive/keras-yolo3/darknet


In [ ]:
!make

mkdir -p obj
mkdir -p backup
mkdir -p results
gcc -Iinclude/ -Isrc/ -Wall -Wno-unused-result -Wno-unknown-pragmas -Wfatal-errors -fPIC -Ofast -c ./src/gemm.c -o obj/gemm.o
gcc -Iinclude/ -Isrc/ -Wall -Wno-unused-result -Wno-unknown-pragmas -Wfatal-errors -fPIC -Ofast -c ./src/utils.c -o obj/utils.o
gcc -Iinclude/ -Isrc/ -Wall -Wno-unused-result -Wno-unknown-pragmas -Wfatal-errors -fPIC -Ofast -c ./src/cuda.c -o obj/cuda.o
gcc -Iinclude/ -Isrc/ -Wall -Wno-unused-result -Wno-unknown-pragmas -Wfatal-errors -fPIC -Ofast -c ./src/deconvolutional_layer.c -o obj/deconvolutional_layer.o
gcc -Iinclude/ -Isrc/ -Wall -Wno-unused-result -Wno-unknown-pragmas -Wfatal-errors -fPIC -Ofast -c ./src/convolutional_layer.c -o obj/convolutional_layer.o
gcc -Iinclude/ -Isrc/ -Wall -Wno-unused-result -Wno-unknown-pragmas -Wfatal-errors -fPIC -Ofast -c ./src/list.c -o obj/list.o
gcc -Iinclude/ -Isrc/ -Wall -Wno-unused-result -Wno-unknown-pragmas -Wfatal-errors -fPIC -Ofast -c ./src/image.c -o obj/image.o
gcc 

In [ ]:
!wget https://pjreddie.com/media/files/yolov3.weights

--2020-09-25 02:56:06--  https://pjreddie.com/media/files/yolov3.weights
Resolving pjreddie.com (pjreddie.com)... 128.208.4.108
Connecting to pjreddie.com (pjreddie.com)|128.208.4.108|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 248007048 (237M) [application/octet-stream]
Saving to: ‘yolov3.weights’

yolov3.weights      100%[===================>] 236.52M   227KB/s    in 11m 28s 

2020-09-25 03:07:35 (352 KB/s) - ‘yolov3.weights’ saved [248007048/248007048]



In [ ]:
!cp yolov3.weights ../

In [ ]:
cd ..

/content/drive/My Drive/keras-yolo3


In [ ]:
!python convert.py yolov3.cfg yolov3.weights model_data/yolo.h5

Using TensorFlow backend.
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:529: FutureWarn

In [ ]:
cd darknet/

/content/drive/My Drive/keras-yolo3/darknet


In [ ]:
!./darknet detector test cfg/coco.data cfg/yolov3.cfg yolov3.weights my_testset/IMG_1843.JPG

layer     filters    size              input                output
    0 conv     32  3 x 3 / 1   608 x 608 x   3   ->   608 x 608 x  32  0.639 BFLOPs
    1 conv     64  3 x 3 / 2   608 x 608 x  32   ->   304 x 304 x  64  3.407 BFLOPs
    2 conv     32  1 x 1 / 1   304 x 304 x  64   ->   304 x 304 x  32  0.379 BFLOPs
    3 conv     64  3 x 3 / 1   304 x 304 x  32   ->   304 x 304 x  64  3.407 BFLOPs
    4 res    1                 304 x 304 x  64   ->   304 x 304 x  64
    5 conv    128  3 x 3 / 2   304 x 304 x  64   ->   152 x 152 x 128  3.407 BFLOPs
    6 conv     64  1 x 1 / 1   152 x 152 x 128   ->   152 x 152 x  64  0.379 BFLOPs
    7 conv    128  3 x 3 / 1   152 x 152 x  64   ->   152 x 152 x 128  3.407 BFLOPs
    8 res    5                 152 x 152 x 128   ->   152 x 152 x 128
    9 conv     64  1 x 1 / 1   152 x 152 x 128   ->   152 x 152 x  64  0.379 BFLOPs
   10 conv    128  3 x 3 / 1   152 x 152 x  64   ->   152 x 152 x 128  3.407 BFLOPs
   11 res    8                 152 x 

In [ ]:
cd ..

/content/drive/My Drive/keras-yolo3


In [ ]:
!python yolo_video.py --image

Using TensorFlow backend.
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:529: FutureWarn

### 【問題4】学習のためのファイルを作成
新しいデータ（シンプソンズデータセット）を学習します。README.mdのTrainingの項を読み、シンプソンズデータセットを学習するために必要なファイルを作成してください。

アノテーションファイルの形式が問題1の実装とは異なるため変換する必要があります。

#### アノテーションファイル

In [ ]:
!head annotation.txt

simpsons_dataset/abraham_grampa_simpson/pic_0000.jpg 57,72,52,72,1
simpsons_dataset/abraham_grampa_simpson/pic_0001.jpg 80,31,337,354,1
simpsons_dataset/abraham_grampa_simpson/pic_0002.jpg 128,48,285,407,1
simpsons_dataset/abraham_grampa_simpson/pic_0003.jpg 72,126,158,275,1
simpsons_dataset/abraham_grampa_simpson/pic_0004.jpg 123,61,294,416,1
simpsons_dataset/abraham_grampa_simpson/pic_0005.jpg 115,18,498,413,1
simpsons_dataset/abraham_grampa_simpson/pic_0006.jpg 171,47,423,413,1
simpsons_dataset/abraham_grampa_simpson/pic_0007.jpg 120,53,381,409,1
simpsons_dataset/abraham_grampa_simpson/pic_0008.jpg 149,56,398,406,1
simpsons_dataset/abraham_grampa_simpson/pic_0009.jpg 205,41,470,456,1


#### クラスファイル

In [ ]:
!cat model_data/Simpsons_classes.txt

abraham_grampa_simpson
apu_nahasapeemapetilon
bart_simpson
charles_montgomery_burns
chief_wiggum
comic_book_guy
edna_krabappel
homer_simpson
kent_brockman
krusty_the_clown
lisa_simpson
marge_simpson
milhouse_van_houten
moe_szyslak
ned_flanders
nelson_muntz
principal_skinner
sideshow_bob



### 【問題5】学習が行えることの確認
問題4で作成したファイルを使用して学習してください。実行環境で学習に時間がかかる場合は、学習が行えることを確認するのみで終えて構いません。

ローカル環境で学習できることを確認済み。
#### 環境
- tensorflow-gpu:1.13.1
- keras:2.1.5


実行結果抜粋  

I tensorflow/stream_executor/dso_loader.cc:152] successfully opened CUDA library libcublas.so.10.0 locally
221/221 [==============================] - 381s 2s/step - loss: 868.9532 - val_loss: 4905.9934
Epoch 2/50
221/221 [==============================] - 374s 2s/step - loss: 90.9763 - val_loss: 4639.4459
Epoch 3/50
221/221 [==============================] - 382s 2s/step - loss: 51.4328 - val_loss: 4501.4463
Epoch 4/50
221/221 [==============================] - 414s 2s/step - loss: 37.2070 - val_loss: 4425.5335
Epoch 5/50
221/221 [==============================] - 397s 2s/step - loss: 30.3154 - val_loss: 4334.5240
Epoch 6/50
221/221 [==============================] - 390s 2s/step - loss: 26.5346 - val_loss: 4292.0171
Epoch 7/50
221/221 [==============================] - 387s 2s/step - loss: 23.8764 - val_loss: 4247.5747
Epoch 8/50
221/221 [==============================] - 378s 2s/step - loss: 22.0989 - val_loss: 4232.1254
Epoch 9/50
221/221 [==============================] - 370s 2s/step - loss: 20.9360 - val_loss: 4192.7829
Epoch 10/50
221/221 [==============================] - 374s 2s/step - loss: 19.9809 - val_loss: 4169.2239
Epoch 11/50
221/221 [==============================] - 367s 2s/step - loss: 19.1545 - val_loss: 4130.3175
Epoch 12/50
221/221 [==============================] - 352s 2s/step - loss: 18.7240 - val_loss: 4131.9996
Epoch 13/50
221/221 [==============================] - 341s 2s/step - loss: 17.9371 - val_loss: 4114.0710
Epoch 14/50
221/221 [==============================] - 333s 2s/step - loss: 17.6341 - val_loss: 4096.6797
Epoch 15/50
221/221 [==============================] - 332s 2s/step - loss: 17.3372 - val_loss: 4082.8572
Epoch 16/50
221/221 [==============================] - 331s 1s/step - loss: 17.0722 - val_loss: 4071.4631
Epoch 17/50
221/221 [==============================] - 333s 2s/step - loss: 16.8318 - val_loss: 4063.7292
Epoch 18/50
221/221 [==============================] - 331s 1s/step - loss: 16.6210 - val_loss: 4049.4513
Epoch 19/50
221/221 [==============================] - 332s 2s/step - loss: 16.4712 - val_loss: 4044.3006
Epoch 20/50
221/221 [==============================] - 332s 2s/step - loss: 16.2939 - val_loss: 4029.5608
Epoch 21/50
221/221 [==============================] - 331s 1s/step - loss: 16.0799 - val_loss: 4043.2665
Epoch 22/50
221/221 [==============================] - 332s 2s/step - loss: 15.9311 - val_loss: 4018.7090
Epoch 23/50
221/221 [==============================] - 332s 2s/step - loss: 15.9284 - val_loss: 4019.8314
Epoch 24/50
221/221 [==============================] - 331s 1s/step - loss: 15.9293 - val_loss: 4018.1681
Epoch 25/50
221/221 [==============================] - 331s 1s/step - loss: 15.7598 - val_loss: 4027.2567
Epoch 26/50
221/221 [==============================] - 332s 2s/step - loss: 15.6154 - val_loss: 4000.8113
Epoch 27/50
221/221 [==============================] - 330s 1s/step - loss: 15.5457 - val_loss: 4016.4499
Epoch 28/50
221/221 [==============================] - 330s 1s/step - loss: 15.6155 - val_loss: 3990.1790
Epoch 29/50
221/221 [==============================] - 331s 1s/step - loss: 15.5101 - val_loss: 3983.1874
Epoch 30/50
221/221 [==============================] - 332s 2s/step - loss: 15.4518 - val_loss: 3976.1444
Epoch 31/50
221/221 [==============================] - 332s 2s/step - loss: 15.4708 - val_loss: 3986.0516
Epoch 32/50
221/221 [==============================] - 331s 1s/step - loss: 15.4070 - val_loss: 3968.0175
Epoch 33/50
221/221 [==============================] - 331s 1s/step - loss: 15.3548 - val_loss: 3946.5910
Epoch 34/50
221/221 [==============================] - 333s 2s/step - loss: 15.2946 - val_loss: 3947.4556
Epoch 35/50
221/221 [==============================] - 330s 1s/step - loss: 15.3423 - val_loss: 3944.6254
Epoch 36/50
221/221 [==============================] - 331s 1s/step - loss: 15.2567 - val_loss: 3942.0856
Epoch 37/50
221/221 [==============================] - 331s 1s/step - loss: 15.2830 - val_loss: 3931.9269
Epoch 38/50
221/221 [==============================] - 333s 2s/step - loss: 15.2237 - val_loss: 3918.5342
Epoch 39/50
221/221 [==============================] - 332s 2s/step - loss: 15.2836 - val_loss: 3912.9420
Epoch 40/50
221/221 [==============================] - 334s 2s/step - loss: 15.1752 - val_loss: 3909.9289
Epoch 41/50
221/221 [==============================] - 331s 1s/step - loss: 15.2143 - val_loss: 3902.3931
Epoch 42/50
221/221 [==============================] - 331s 1s/step - loss: 15.1757 - val_loss: 3892.0640
Epoch 43/50
221/221 [==============================] - 332s 2s/step - loss: 15.2077 - val_loss: 3876.6679
Epoch 44/50
221/221 [==============================] - 331s 1s/step - loss: 15.1717 - val_loss: 3885.1287
Epoch 45/50
221/221 [==============================] - 331s 1s/step - loss: 15.1554 - val_loss: 3891.7449
Epoch 46/50
221/221 [==============================] - 331s 1s/step - loss: 15.1819 - val_loss: 3894.2237
Epoch 47/50
221/221 [==============================] - 332s 2s/step - loss: 15.1510 - val_loss: 3924.6045
Epoch 48/50
221/221 [==============================] - 331s 1s/step - loss: 15.1085 - val_loss: 3933.0512
Epoch 49/50
221/221 [==============================] - 331s 1s/step - loss: 15.1736 - val_loss: 3948.5245
Epoch 50/50
221/221 [==============================] - 330s 1s/step - loss: 15.0981 - val_loss: 3970.1722
